Custom array type provokes panic

The code below provokes a panic. If I revert the changes done in https://github.com/gorilla/schema/commit/e6c82218a8b3ed3cbeb5407429849c0b0b597d40 / https://github.com/gorilla/schema/pull/91 the code below works.

I am not sure what exactly how the problem can be solved. I tried to tweak the fix in https://github.com/gorilla/schema/commit/e6c82218a8b3ed3cbeb5407429849c0b0b597d40  but I didn't make too much progress.
```
func TestRegisterEncoderCustomArrayType(t *testing.T) {
	vals := map[string][]string{}
	
	type CustomInt []int
	s1 := &struct {
		SomeInts CustomInt `schema:",omitempty"`
	}{
		CustomInt{1,2,3},
	}
	
	encoder := NewEncoder()
	encoder.RegisterEncoder(CustomInt{}, func(value reflect.Value) string {
		return fmt.Sprint(value.Interface())
	})
	
	encoder.Encode(s1, vals)
	t.Log(vals)
}
```
```
panic: runtime error: comparing uncomparable type schema.CustomInt [recovered]
	panic: runtime error: comparing uncomparable type schema.CustomInt

goroutine 5 [running]:
testing.tRunner.func1(0xc4200a60f0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/testing.go:711 +0x2d2
panic(0x114d340, 0xc420048720)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/panic.go:491 +0x283
github.com/gorilla/schema.(*Encoder).encode(0xc42006bc18, 0x113cc20, 0xc42000a0c0, 0x16, 0xc420074390, 0xc42000a0e0, 0xc420044d28)
	/Users/olve/go/src/github.com/gorilla/schema/encoder.go:70 +0x78b
github.com/gorilla/schema.(*Encoder).Encode(0xc420044c18, 0x113cc20, 0xc42000a0c0, 0xc420074390, 0xc420044c60, 0x56600000002)
	/Users/olve/go/src/github.com/gorilla/schema/encoder.go:29 +0x6b
github.com/gorilla/schema.TestRegisterEncoderCustomArrayType(0xc4200a60f0)
	/Users/olve/go/src/github.com/gorilla/schema/encoder_test.go:352 +0x3a9
testing.tRunner(0xc4200a60f0, 0x117f3a0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/testing.go:789 +0x2de
```

--- a/encoder.go
+++ b/encoder.go
@@ -40,6 +40,29 @@ func (e *Encoder) SetAliasTag(tag string) {
 	e.cache.tag = tag
 }
 
+func isZero(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Func:
+	case reflect.Map, reflect.Slice:
+		return v.IsNil() || v.Len() == 0
+	case reflect.Array:
+		z := true
+		for i := 0; i < v.Len(); i++ {
+			z = z && isZero(v.Index(i))
+		}
+		return z
+	case reflect.Struct:
+		z := true
+		for i := 0; i < v.NumField(); i++ {
+			z = z && isZero(v.Field(i))
+		}
+		return z
+	}
+	// Compare other types directly:
+	z := reflect.Zero(v.Type())
+	return v.Interface() == z.Interface()
+}
+
 func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 	if v.Kind() == reflect.Ptr {
 		v = v.Elem()
@@ -67,7 +90,7 @@ func (e *Encoder) encode(v reflect.Value, dst map[string][]string) error {
 		// Encode non-slice types and custom implementations immediately.
 		if encFunc != nil {
 			value := encFunc(v.Field(i))
-			if opts.Contains("omitempty") && v.Field(i).Interface() == reflect.Zero(reflect.TypeOf(v.Field(i).Interface())).Interface() {
+			if opts.Contains("omitempty") && isZero(v.Field(i)) {
 				continue
 			}
 

--- a/encoder_test.go
+++ b/encoder_test.go
@@ -1,6 +1,7 @@
 package schema
 
 import (
+	"fmt"
 	"reflect"
 	"testing"
 )
@@ -329,4 +330,29 @@ func TestEncoderWithOmitempty(t *testing.T) {
 	valExists(t, "f0601", "test", vals)
 	valNotExists(t, "f08", vals)
 	valsExist(t, "f09", []string{"test"}, vals)
+}
+
+func TestRegisterEncoderCustomArrayType(t *testing.T) {
+	type CustomInt []int
+	type S1 struct {
+		SomeInts CustomInt `schema:",omitempty"`
+	}
+
+	ss := []S1{
+		{},
+		{CustomInt{}},
+		{CustomInt{1, 2, 3}},
+	}
+
+	for s := range ss {
+		vals := map[string][]string{}
+
+		encoder := NewEncoder()
+		encoder.RegisterEncoder(CustomInt{}, func(value reflect.Value) string {
+			return fmt.Sprint(value.Interface())
+		})
+
+		encoder.Encode(s, vals)
+		t.Log(vals)
+	}
 }

